docs/architecture/deployment-execution/steps.md

```diff
@@ -0,0 +1,257 @@
+# Deployment Steps
+
+Deployments in Pulumi happens as a series of `Step`s. These are generated by the
+[step generator](https://pulumi-developer-docs.readthedocs.io/latest/docs/architecture/deployment-execution/resource-registration.html#step-generation),
+and then executed by the deployment executor.
+
+The main functionality of steps is in the `Apply` function. that is called,
+the step is executed, which means the provider is called as appropriate
+(depending on the type of step we're executing), and a completion function is
+returned, which when called registers the result of the step.
+
+Each step also has an old and a new state, the old state representing the state
+we read from the snapshot, and the new state representing the state that the
+step should have after the `Apply()`.
+
+Next we're going to describe the steps in a little more detail.
+
+## Overview
+
+Pulumi's deployment system uses different step types to represent the various
+operations that can be performed on resources. These steps fall into several
+categories:
+
+### Step Categories
+
+**Mutating Steps** directly modify resources via provider operations:
```

Comment:
Mutating is confusing because it's not clear what is being mutated here. Maybe I'd use "Externally mutating"?